source/h2_decoder.c

```diff
@@ -483,61 +607,32 @@ static int s_state_fn_priority_block(struct aws_h2_decoder *decoder, struct aws_
      * https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-9513
      */
 
-    const enum aws_h2_frame_type frame_type = decoder->frame_in_progress.type;
-    s_decoder_set_state(decoder, s_state_frames[frame_type]);
+    s_decoder_go_to_frame_state(decoder);
 
     return AWS_OP_SUCCESS;
 }
 
 static int s_state_fn_frame_data(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
 
-    const uint32_t remaining_length = decoder->frame_in_progress.payload_len;
-
-    struct aws_byte_cursor body_to_pass;
-    bool will_finish_state;
-    if (input->len < remaining_length) {
-        body_to_pass = aws_byte_cursor_advance(input, input->len);
-        will_finish_state = false;
-    } else {
-        body_to_pass = aws_byte_cursor_advance(input, remaining_length);
-        will_finish_state = true;
+    const struct aws_byte_cursor body_data = s_decoder_get_payload(decoder, input);
+    if (body_data.len) {
+        DECODER_CALL_VTABLE_STREAM_ARGS(decoder, on_data, &body_data);
     }
-    AWS_FATAL_ASSERT(body_to_pass.len <= UINT32_MAX);
-
-    /* Update the payload len to just be what's left */
-    decoder->frame_in_progress.payload_len -= (uint32_t)body_to_pass.len;
 
-    DECODER_CALL_VTABLE_STREAM_ARGS(decoder, on_data, &body_to_pass);
-
-    if (will_finish_state) {
+    if (decoder->frame_in_progress.payload_len == 0) {
         /* Process padding if necessary, otherwise we're done! */
-        if (decoder->frame_in_progress.flags & AWS_H2_FRAME_F_PADDED) {
-            s_decoder_set_state(decoder, &s_state_padding);
-        } else {
-            s_decoder_reset_state(decoder);
-        }
+        s_decoder_set_state(decoder, &s_state_padding);
```

Comment:
We end up waiting for MORE data before we can finish processing THIS data
fix it colden fix it